### PR TITLE
Properly consider APITs for never type fallback ascription fix

### LINKED ISSUE #132853: 2024 impl_trait_overcaptures ambiguous `+`
I tried this code:

```rust
#![warn(impl_trait_overcaptures)]

pub fn f(x: &i32) -> &impl Clone {
    x
}
```

I expected to see this happen: Provides a valid suggestion that is `-> &(impl Clone + use<>)` with parentheses to avoid ambiguity.

Instead, this happened: Provides a suggestion that includes `&impl Clone + use<>` which fails to parse with:

```
error: ambiguous `+` in a type
 --> src/main.rs:3:23
  |
3 | pub fn f(x: &i32) -> &impl Clone + use<> {
  |                       ^^^^^^^^^^^^^^^^^^
  |
help: try adding parentheses
  |
3 | pub fn f(x: &i32) -> &(impl Clone + use<>) {
  |                       +                  +
```

### Meta

`rustc --version --verbose`:
```
rustc 1.84.0-nightly (b91a3a056 2024-11-07)
binary: rustc
commit-hash: b91a3a05609a46f73d23e0995ae7ebb4a4f429a5
commit-date: 2024-11-07
host: aarch64-apple-darwin
release: 1.84.0-nightly
LLVM version: 19.1.3
```

### LINKED ISSUE #133842: dependency_on_unit_never_type_fallback bad suggestion with APIT generics
The `dependency_on_unit_never_type_fallback` lint seems to provide a bad suggestion to include too many generic arguments when the function being called uses APIT. Example:

```rust
#![warn(rust_2024_compatibility)]

pub fn foo<T: Default>(_y: impl Fn()) -> Result<T, ()> {
    Err(())
}

pub fn bar() -> Result<(), ()> {
    foo(|| {})?;
    Ok(())
}
```

Here this suggests the following change:

```diff
@@ -5,6 +5,6 @@
 }
 
 pub fn bar() -> Result<(), ()> {
-    foo(|| {})?;
+    foo::<(), _>(|| {})?;
     Ok(())
 }
```

However, this fails to compile because the function only takes one generic argument. The correct suggestion would be just `::<()>`.

### Meta

```
rustc 1.85.0-nightly (c44b3d50f 2024-12-03)
binary: rustc
commit-hash: c44b3d50fea96a3e0417e8264c16ea21a0a3fca2
commit-date: 2024-12-03
host: aarch64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.4
```

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #133842

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? WaffleLapkin or maybe re-roll if you're busy :>

File: compiler/rustc_hir_typeck/src/fallback.rs [hunk 1]
```diff
@@ -613,17 +613,21 @@ impl<'tcx> AnnotateUnitFallbackVisitor<'_, 'tcx> {
...
+                })
                 .count();
-            for (idx, arg) in args
-                .iter()
-                .filter(|arg| matches!(arg.unpack(), ty::GenericArgKind::Type(_)))
+            for (idx, (arg, _)) in args
+                .filter(|(_, param)| {
+                    matches!(param.kind, ty::GenericParamDefKind::Type { synthetic: false, .. })
+                })
                 .enumerate()
```
Comment (WaffleLapkin) [inline compiler/rustc_hir_typeck/src/fallback.rs:631]: Can you create an iterator and then clone it instead? i.e. `let args = args.filter()` and then `args.clone().count()` + `for ... in args.enumarate()`.

**Review by WaffleLapkin — APPROVED** [review]
r=me after minor nit

File: compiler/rustc_hir_typeck/src/fallback.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/fallback.rs:631]: good point, can deduplicate the filtering

**Pull request merged**